### PR TITLE
yarn: fix CWE parsing

### PR DESCRIPTION
**Pull request by valentijnscholten**
fix yarn parser to handle wildly formatted list of CWEs: #5999

        # Somehow multiple CWEs end up in the json report like this
        # [\"CWE-173\",\"CWE-200\",\"CWE-601\"]
        # which becomes after json load:
        # ["CWE-173","CWE-200","CWE-601"]
        # we parse this and take the first CWE

EDIT: Seems to happen also for single CWE values

        # [\"CWE-173\"]

### CODE REVIEW

**Comment by dvdgsng** [comment]
Any idea what DefectDojo is doing when running an import that could cause such an issue out of the blue? We are running v2.7.1 and haven't touched the deployment for 25 days. So we would expect it to be rather stable.

**Comment by valentijnscholten** [comment]
yarn has changed their output format or logic, or gets data from another source formatted in a different way.

**Comment by dvdgsng** [comment]
AFAICT we haven't changed the node version either :/ need to verify though

**Comment by valentijnscholten** [comment]
It might be a bug / side effect in yarn that they may fix at some point.

File: dojo/tools/yarn_audit/parser.py [hunk 1]
```diff
@@ -54,6 +55,25 @@ def get_item(item_node, test):
...
+        # we parse this and take the first CWE
+
+        cwe = ast.literal_eval(cwe)[0][4:]
+        # https://docs.python.org/3/library/ast.html#ast.literal_eval
+        # Safely evaluate an expression node or a string containing a Python literal or container display.
+        # The string or node provided may only consist of the following Python literal structures:
+        # strings, bytes, numbers, tuples, lists, dicts, sets, booleans, None and Ellipsis.
+        # This can be used for safely evaluating strings containing Python values from untrusted sources
+        #  without the need to parse the values oneself. It is not capable of evaluating arbitrarily
+        #  complex expressions, for example involving operators or indexing.
```
Comment (damiencarol) [inline dojo/tools/yarn_audit/parser.py:75]: I think it's better to use `json.loads()`. The data in the report is not Python code.
It's JSON object in a JSON string. so better to load it this way (with the JSON module)

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: dojo/tools/yarn_audit/parser.py [hunk 2]
```diff
@@ -54,6 +54,18 @@ def get_item(item_node, test):
...
+    if cwe.startswith('CWE-'):
+        cwe = int(cwe[4:])
+    elif cwe.startswith('['):
+        # Somehow multiple CWEs end up in the json report like this
+        # [\"CWE-173\",\"CWE-200\",\"CWE-601\"]
+        # which becomes after json load:
+        # ["CWE-173","CWE-200","CWE-601"]
+        # we parse this and take the first CWE
+
+        cwe = json.loads(cwe)[0][4:]
```
Comment (damiencarol) [inline dojo/tools/yarn_audit/parser.py:67]:
```suggestion
        cwe = json.loads(cwe)[0]
```
Should it be like this?

Comment (valentijnscholten) [reply dojo/tools/yarn_audit/parser.py:67]: No, cwe is an integer field.

**Review by damiencarol — APPROVED** [review]

**Review by devGregA — APPROVED** [review]

**Pull request merged**